custom_metrics/almanac.js

```diff
@@ -12,32 +12,263 @@
+    var schemaElements = {};
+
+    if (nodes) {
+      for (var i = 0, len = nodes.length; i < len; i++) {
+        var node = nodes[i];
+        var item = node.getAttribute('itemtype');
+
+        if (item) {
+          // microdata
+          link.href = node.getAttribute('itemtype');
+          schemaElements[link.hostname + link.pathname] = true;
+        } else if ((node.tagName = 'SCRIPT')) {
+          // json+ld
+          var content = JSON.parse(node.textContent);
+          var contentLoop = [];
+          if (content) {
+            // nested lookup
+            nestedLookup(content);
+          }
+        }
+      }
+    }
+    return Object.keys(schemaElements);
+  })(),
+  // Looks at links and identifies internal, external or hashed
+  'seo-anchor-elements': (() => {
+    // metric 10.10, 10.11,
+    var nodes = document.getElementsByTagName('a');
+    var link = document.createElement('a');
+
+    var internal = 0; // metric 10.10
+    var external = 0; // metric 10.10
+    var hash = 0;
+    var navigateHash = 0; // metric 10.11
 
-            for (var n = 0, len2 = attributes.length; n < len2; n++) {
-                var attribute = attributes[n];
-                el[attribute.name.toLowerCase()] = attribute.value;
+    if (nodes) {
+      for (var i = 0, len = nodes.length; i < len; i++) {
+        var node = nodes[i];
+
+        // our local parser trick
+        if (node.href) {
+          link.href = node.href;
+
+          if (document.location.hostname === link.hostname) {
+            internal++;
+            if (document.location.pathname === link.pathname && link.hash.length > 0) {
+              // check if hash matches an element in the DOM (scroll to)
+              var element = document.querySelector(link.hash);
+              if (element) {
+                hash++;
+              } else {
+                navigateHash++;
+              }
+            }
+          } else if (document.location.hostname !== link.hostname) {
+            external++;
+          }
+        }
+      }
+    }
+
+    return { internal, external, hash, navigateHash };
+  })(),
+  // Extracts titles used and counts the words, to flag thin content pages
+  'seo-titles': (() => {
+    //metric 10.9
+    var nodes = document.querySelectorAll('h1, h2, h3, h4');
+    var titleWords = -1;
+    var titleElements = -1;
+
+    if (nodes) {
+      titleWords = 0;
+      titleElements = 0;
+
+      // analyse each title node
+      for (var i = 0, len = nodes.length; i < len; i++) {
+        var node = nodes[i];
+        // remove extra whitespace
+        var nodeText = node.textContent.replace(/\s+/g, ' ').replace(/^\s+|\s+$/, '');
+        var nodeWordsCount = nodeText.split(' ').length;
+
+        if (nodeWordsCount > 0) {
+          titleWords += nodeWordsCount;
+          titleElements++;
+        }
+      }
+    }
+    return { titleWords, titleElements };
+  })(),
+  // Extracts words on the page to flag thin content pages
```

Comment:
Is there a significant performance cost to iterating over all nodes in the DOM? For some text-heavy pages, I also wonder if saving everything to an array could eat up significant memory to affect test performance/accuracy.

Reply:
About performance, I'm not sure right person to answer. This is DOM native method.

I've removed the array loop and replaced it with a function call.

Reply:
My concern was mostly about how often the `analyseNode` function would be called per page. Any clue whether it might be called 100 times? 10,000 times? 1,000,000 times?

Reply:
analyseNode is only applied to text nodes. I tested on Wikipedia's largest articles:
> https://en.wikipedia.org/wiki/List_of_2018%E2%80%9319_Super_Rugby_transfers = 21,568 times
> https://en.wikipedia.org/wiki/List_of_2018%E2%80%9319_Super_Rugby_transfers = 24770 times